teps/0121-unified-retries-for-runtime-objects.md

```diff
@@ -0,0 +1,421 @@
+
+| Runtime Object | Retriable Independently? | Retriable in Pipeline? | Timeout for Per-retry? | Note |
+|:---|:---:|:---:|:---:|:---:|
+|Task Run| No|Yes| Yes| The retry logic lives in PipelineRun controller|
+|Custom Run| Yes| Yes | No| The retry logic is in the custom task controller|
+
+Another worth-mentioning detail is we retry a TaskRun by moving its current status to retriesStatus, clearing up timestamps like startTime, completionTime. (Another possible way is recreating another TaskRun instead of updating existing TaskRun in-place.)
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this TEP. Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports][experience reports]
+to demonstrate the interest in a TEP within the wider Tekton community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Unify API spec for `retries` and `timeout`
+- Unify implementation pattern.
+<!--
+List the specific goals of the TEP.
+- What is it trying to achieve?
+- How will we know that this has succeeded?
+-->
+
+### Non-Goals
+- Configurable retry strategy
+
+<!--
+Listing non-goals helps to focus discussion and make progress.
+- What is out of scope for this TEP?
+-->
+
+### Use Cases
+
+`Retry` is a common functionality across all different systems. The blog post "[Transient faults handling](https://docs.microsoft.com/en-us/azure/architecture/best-practices/transient-faults)" from Microsoft Azure clearly states what are `Transient faults`, provides a general guideline of designing a successful `retry strategy`. 
+
+Quote the summary here:
+> All applications that communicate with remote services and resources must be sensitive to transient faults. This is especially the case for applications that run in the cloud, where the nature of the environment and connectivity over the Internet means these types of faults are likely to be encountered more often. Transient faults include the momentary loss of network connectivity to components and services, the temporary unavailability of a service, or timeouts that arise when a service is busy. These faults are often self-correcting, and if the action is repeated after a suitable delay it is likely to succeed.
+
+It lists several use cases:
+- the momentary loss of network connectivity to components and services
+- the temporary unavailability of a service
+- timeouts that arise when a service is busy
+
+Which all make sense to us as a Cloud Native CI/CD system.
+
+### Requirements
+
+- The behavior of retry should be unified across all runtime objects including TaskRun, CustomRun, potentially as well as PipelineRun once [TEP-0056](https://github.com/tektoncd/community/blob/main/teps/0056-pipelines-in-pipelines.md) is implemented.
+- The implement pattern of the `retry` functionality should be unified across all run time objects.
```

Comment:
```suggestion
- The implementation pattern of the `retry` functionality should be unified across all run time objects.
```

Reply:
for Runs, we can only provide a recommendation that aligns with what we do in TaskRuns, but believe we should leave it up to the users to implement a retry functionality that works for their use cases - this was discussed in https://github.com/tektoncd/community/blob/main/teps/0069-support-retries-for-custom-task-in-a-pipeline.md

Reply:
Thanks for the comment! That's also what I thought but the meaning it expresses seems misleading lol, I rephrased it to reflect the idea.